File: services/driver-app/src/utils/locationValidation.js
/**
 * Distance validation utility for pickup/delivery
 * Ensures driver is within acceptable range of source/destination
 */

const EARTH_RADIUS_KM = 6371;

/**
 * Calculate distance between two coordinates using Haversine formula
 * Returns distance in kilometers
 */
export function calculateDistance(lat1, lon1, lat2, lon2) {
  const toRad = (deg) => (deg * Math.PI) / 180;

  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) *
      Math.cos(toRad(lat2)) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_KM * c;
}

/**
 * Check if driver is within acceptable range of location
 * @param {object} driverLocation - { latitude, longitude }
 * @param {object} targetLocation - { latitude, longitude }
 * @param {number} radiusKm - Acceptable radius in kilometers (default 50km)
 */
export function isWithinLocation(driverLocation, targetLocation, radiusKm = 50) {
  if (!driverLocation || !targetLocation) {
    return false;
  }

  const distance = calculateDistance(
    driverLocation.latitude,
    driverLocation.longitude,
    targetLocation.latitude,
    targetLocation.longitude
  );

  return distance <= radiusKm;
}

/**
 * Get distance and direction from driver to target location
 */
export function getLocationInfo(driverLocation, targetLocation) {
  if (!driverLocation || !targetLocation) {
    return { distance: null, direction: null };
  }

  const distance = calculateDistance(
    driverLocation.latitude,
    driverLocation.longitude,
    targetLocation.latitude,
    targetLocation.longitude
  );

  // Calculate bearing (direction)
  const lat1 = driverLocation.latitude;
  const lon1 = driverLocation.longitude;
  const lat2 = targetLocation.latitude;
  const lon2 = targetLocation.longitude;

  const dLon = lon2 - lon1;
  const y = Math.sin(dLon) * Math.cos(lat2);
  const x =
    Math.cos(lat1) * Math.sin(lat2) -
    Math.sin(lat1) * Math.cos(lat2) * Math.cos(dLon);

  const bearing = (Math.atan2(y, x) * 180) / Math.PI;
  const direction = ((bearing + 360) % 360).toFixed(0);

  return {
    distance: distance.toFixed(2),
    direction: `${direction}°`,
  };
}

/**
 * Get human-readable distance description
 */
export function getDistanceDescription(distanceKm) {
  if (distanceKm < 0.1) {
    return 'You are at the location';
  }
  if (distanceKm < 0.5) {
    return `${(distanceKm * 1000).toFixed(0)}m away`;
  }
  if (distanceKm < 2) {
    return `${distanceKm.toFixed(1)} km away`;
  }
  return `${distanceKm.toFixed(1)} km away - too far`;
}

File: services/backend/internal/utils/location_validation.go
package utils

import (
	"fmt"
	"math"
)

const EarthRadiusKm = 6371.0

// LocationValidation provides location-based validation utilities
type LocationValidation struct {
	SourceLat       float64
	SourceLon       float64
	DestLat         float64
	DestLon         float64
	AllowedRadiusKm float64
}

// NewLocationValidation creates a new LocationValidation instance
func NewLocationValidation(srcLat, srcLon, dstLat, dstLon, radiusKm float64) *LocationValidation {
	return &LocationValidation{
		SourceLat:       srcLat,
		SourceLon:       srcLon,
		DestLat:         dstLat,
		DestLon:         dstLon,
		AllowedRadiusKm: radiusKm,
	}
}

// CalculateDistance calculates distance between two coordinates using Haversine formula
// Returns distance in kilometers
func CalculateDistance(lat1, lon1, lat2, lon2 float64) float64 {
	toRad := func(deg float64) float64 { return deg * math.Pi / 180.0 }

	dLat := toRad(lat2 - lat1)
	dLon := toRad(lon2 - lon1)

	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(toRad(lat1))*math.Cos(toRad(lat2))*math.Sin(dLon/2)*math.Sin(dLon/2)

	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	return EarthRadiusKm * c
}

// IsWithinSourceLocation checks if driver is within allowed radius of source
func (lv *LocationValidation) IsWithinSourceLocation(driverLat, driverLon float64) bool {
	distance := CalculateDistance(driverLat, driverLon, lv.SourceLat, lv.SourceLon)
	return distance <= lv.AllowedRadiusKm
}

// IsWithinDestinationLocation checks if driver is within allowed radius of destination
func (lv *LocationValidation) IsWithinDestinationLocation(driverLat, driverLon float64) bool {
	distance := CalculateDistance(driverLat, driverLon, lv.DestLat, lv.DestLon)
	return distance <= lv.AllowedRadiusKm
}

// GetDistanceToSource returns distance to source in km
func (lv *LocationValidation) GetDistanceToSource(driverLat, driverLon float64) float64 {
	return CalculateDistance(driverLat, driverLon, lv.SourceLat, lv.SourceLon)
}

// GetDistanceToDestination returns distance to destination in km
func (lv *LocationValidation) GetDistanceToDestination(driverLat, driverLon float64) float64 {
	return CalculateDistance(driverLat, driverLon, lv.DestLat, lv.DestLon)
}

// ValidatePickup validates pickup location and returns error if invalid
func (lv *LocationValidation) ValidatePickup(driverLat, driverLon float64) error {
	if !lv.IsWithinSourceLocation(driverLat, driverLon) {
		distance := lv.GetDistanceToSource(driverLat, driverLon)
		return fmt.Errorf("pickup not allowed: driver is %.2f km from source (allowed: %.1f km)", distance, lv.AllowedRadiusKm)
	}
	return nil
}

// ValidateDelivery validates delivery location and returns error if invalid
func (lv *LocationValidation) ValidateDelivery(driverLat, driverLon float64) error {
	if !lv.IsWithinDestinationLocation(driverLat, driverLon) {
		distance := lv.GetDistanceToDestination(driverLat, driverLon)
		return fmt.Errorf("delivery not allowed: driver is %.2f km from destination (allowed: %.1f km)", distance, lv.AllowedRadiusKm)
	}
	return nil
}
